File: js/managed/src/goals.ts
import { GOAL_CONTINUATION_TEMPLATE } from "./goal-continuation";
import { renderCodexTemplate } from "./codex-prompts";
/** Persisted, session-local goal state. User controls and model tools are deliberately separate. */
export type GoalStatus = "active" | "paused" | "blocked" | "usageLimited" | "budgetLimited" | "complete";
export interface ThreadGoal {
  goalId: string;
  threadId: string;
  objective: string;
  status: GoalStatus;
  tokenBudget?: number;
  tokensUsed: number;
  timeUsedSeconds: number;
  createdAt: number;
  updatedAt: number;
}
export type GoalResponse = { goal: ThreadGoal | null; remainingTokens: number | null; completionBudgetReport: string | null };
export type GoalUserUpdate = { objective?: string; status?: GoalStatus; tokenBudget?: number | null };

function objective(value: unknown): string {
  if (typeof value !== "string" || !value.trim() || [...value.trim()].length > 4_000) {
    throw new TypeError("goal objective must contain 1–4000 characters");
  }
  return value.trim();
}
function budget(value: unknown): number {
  if (typeof value !== "number" || !Number.isSafeInteger(value) || value <= 0) throw new TypeError("token budget must be a positive safe integer");
  return value;
}
export function goalResponse(goal: ThreadGoal | null, completion = false): GoalResponse {
  return { goal, remainingTokens: goal?.tokenBudget === undefined ? null : Math.max(0, goal.tokenBudget - goal.tokensUsed),
    completionBudgetReport: completion && goal?.tokenBudget !== undefined
      ? `Goal completed using ${goal.tokensUsed} of ${goal.tokenBudget} budgeted tokens.` : null };
}

export class Goals {
  constructor(private readonly storage: DurableObjectStorage, private readonly threadId: () => string, private readonly now = Date.now) {
    storage.sql.exec(`CREATE TABLE IF NOT EXISTS managed_goal (singleton INTEGER PRIMARY KEY CHECK(singleton=1), body TEXT NOT NULL)`);
    storage.sql.exec(`CREATE TABLE IF NOT EXISTS managed_goal_usage (goal_id TEXT NOT NULL, turn_id TEXT NOT NULL, tokens INTEGER NOT NULL, seconds REAL NOT NULL, PRIMARY KEY(goal_id,turn_id))`);
  }
  get(): ThreadGoal | null {
    const row = this.storage.sql.exec<{ body: string }>("SELECT body FROM managed_goal WHERE singleton=1").toArray()[0];
    return row ? JSON.parse(row.body) as ThreadGoal : null;
  }
  private save(goal: ThreadGoal): ThreadGoal {
    this.storage.sql.exec("INSERT INTO managed_goal(singleton,body) VALUES(1,?) ON CONFLICT(singleton) DO UPDATE SET body=excluded.body", JSON.stringify(goal));
    return goal;
  }
  create(input: { objective: string; token_budget?: number }): ThreadGoal {
    const text = objective(input.objective);
    const tokenBudget = input.token_budget === undefined ? undefined : budget(input.token_budget);
    return this.storage.transactionSync(() => {
      const current = this.get();
      if (current && current.status !== "complete") throw new Error("cannot create a new goal because this thread has an unfinished goal");
      this.storage.sql.exec("DELETE FROM managed_goal_usage");
      const now = this.now();
      return this.save({ goalId: crypto.randomUUID(), threadId: this.threadId(), objective: text, status: "active", tokenBudget,
        tokensUsed: 0, timeUsedSeconds: 0, createdAt: now, updatedAt: now });
    });
  }
  /** Only expose to an authenticated user control, never as model tool arguments. */
  updateByUser(update: GoalUserUpdate): ThreadGoal {
    const text = update.objective === undefined ? undefined : objective(update.objective);
    const tokenBudget = update.tokenBudget == null ? update.tokenBudget : budget(update.tokenBudget);
    if (update.status !== undefined && !["active", "paused", "blocked", "usageLimited", "budgetLimited", "complete"].includes(update.status)) throw new TypeError("invalid goal status");
    return this.storage.transactionSync(() => {
      const goal = this.get();
      if (!goal) throw new Error("this thread has no goal");
      if (text !== undefined) goal.objective = text;
      if (tokenBudget !== undefined) goal.tokenBudget = tokenBudget ?? undefined;
      if (update.status !== undefined) goal.status = update.status;
      if (goal.tokenBudget !== undefined && goal.tokensUsed >= goal.tokenBudget && goal.status !== "complete") goal.status = "budgetLimited";
      goal.updatedAt = this.now();
      return this.save(goal);
    });
  }
  updateByModel(status: "complete" | "blocked" | "paused"): ThreadGoal {
    if (!["complete", "blocked", "paused"].includes(status)) throw new TypeError("update_goal can only mark a goal complete, blocked, or paused");
    const goal = this.get();
    if (!goal) throw new Error("this thread has no goal");
    // A model cannot undo a system-enforced stop or resume a stopped goal.
    if (goal.status === "budgetLimited" || goal.status === "usageLimited") return goal;
    return this.updateByUser({ status });
  }
  /** Cumulative snapshots for one turn; replay and increasing partial snapshots are safe.
   * Capture goalId at turn admission, so an old turn cannot charge a replacement goal.
   * Tokens should be uncached input + output tokens (the upstream goal accounting policy).
   */
  accountTurn(goalId: string, turnId: string, tokens: number, seconds: number): ThreadGoal | null {
    if (!Number.isSafeInteger(tokens) || tokens < 0 || !Number.isFinite(seconds) || seconds < 0) throw new TypeError("invalid goal usage");
    return this.storage.transactionSync(() => {
      const goal = this.get();
      if (!goal || goal.goalId !== goalId) return goal;
      const previous = this.storage.sql.exec<{ tokens: number; seconds: number }>("SELECT tokens,seconds FROM managed_goal_usage WHERE goal_id=? AND turn_id=?", goalId, turnId).toArray()[0];
      const nextTokens = Math.max(previous?.tokens ?? 0, tokens);
      const nextSeconds = Math.max(previous?.seconds ?? 0, seconds);
      goal.tokensUsed += nextTokens - (previous?.tokens ?? 0);
      goal.timeUsedSeconds += nextSeconds - (previous?.seconds ?? 0);
      if (!Number.isSafeInteger(goal.tokensUsed)) throw new Error("goal token usage overflow");
      this.storage.sql.exec("INSERT INTO managed_goal_usage VALUES(?,?,?,?) ON CONFLICT(goal_id,turn_id) DO UPDATE SET tokens=excluded.tokens,seconds=excluded.seconds", goalId, turnId, nextTokens, nextSeconds);
      if (goal.tokenBudget !== undefined && goal.tokensUsed >= goal.tokenBudget && goal.status !== "complete") goal.status = "budgetLimited";
      goal.updatedAt = this.now();
      return this.save(goal);
    });
  }
  clear(): void {
    this.storage.transactionSync(() => {
      this.storage.sql.exec("DELETE FROM managed_goal");
      this.storage.sql.exec("DELETE FROM managed_goal_usage");
    });
  }
}

/** Call only after successful turn completion; errors/cancellation must not auto-resume. */
export function goalContinuation(goal: ThreadGoal | null): string | null {
  if (!goal || goal.status !== "active") return null;
  const escaped = goal.objective.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
  const values: Record<string, string> = { objective: escaped, tokens_used: String(goal.tokensUsed),
    token_budget: goal.tokenBudget === undefined ? "none" : String(goal.tokenBudget),
    remaining_tokens: goal.tokenBudget === undefined ? "unbounded" : String(Math.max(0, goal.tokenBudget - goal.tokensUsed)) };
  return renderCodexTemplate(GOAL_CONTINUATION_TEMPLATE, values);
}

File: crates/nanocodex-tools/src/extensions/goals.rs
//! Goal tools over host-owned persistence and live accounting.
//!
//! The host implements this adapter using its thread store and must flush live
//! usage before get/update. No process-global or in-memory pretend goal store.
use super::{ExtensionProvider, error};
use crate::{ToolContext, ToolOutput, ToolResult};
use serde::Deserialize;
use serde_json::Value;
#[async_trait::async_trait]
pub trait GoalStore: Send + Sync {
    async fn get(&self, context: ToolContext<'_>) -> Result<Value, crate::contract::ToolError>;
    async fn create(
        &self,
        objective: String,
        token_budget: Option<u64>,
        context: ToolContext<'_>,
    ) -> Result<Value, crate::contract::ToolError>;
    async fn update(
        &self,
        status: GoalStatus,
        context: ToolContext<'_>,
    ) -> Result<Value, crate::contract::ToolError>;
}
#[derive(Clone, Copy, Debug, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GoalStatus {
    Complete,
    Blocked,
    Paused,
}
pub struct GoalTools<S: GoalStore>(pub S);
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct CreateArgs {
    objective: String,
    token_budget: Option<u64>,
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct UpdateArgs {
    status: GoalStatus,
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct EmptyArgs {}
#[async_trait::async_trait]
impl<S: GoalStore> ExtensionProvider for GoalTools<S> {
    fn names(&self) -> &'static [&'static str] {
        &["get_goal", "create_goal", "update_goal"]
    }
    async fn execute(&self, name: &str, input: Value, context: ToolContext<'_>) -> ToolResult {
        let value = match name {
            "get_goal" => {
                let _: EmptyArgs = serde_json::from_value(input)?;
                self.0.get(context).await?
            }
            "create_goal" => {
                let a: CreateArgs = serde_json::from_value(input)?;
                if a.objective.trim().is_empty()
                    || a.objective.trim().chars().count() > 4000
                    || a.token_budget == Some(0)
                {
                    return Err(error("invalid goal objective or token budget"));
                }
                self.0
                    .create(a.objective.trim().to_owned(), a.token_budget, context)
                    .await?
            }
            "update_goal" => {
                let a: UpdateArgs = serde_json::from_value(input)?;
                self.0.update(a.status, context).await?
            }
            _ => return Err(error("unknown goal tool")),
        };
        Ok(ToolOutput::json(&value))
    }
}

use serde::Serialize;
use serde_json::json;
use std::{
    collections::BTreeMap,
    path::PathBuf,
    time::{SystemTime, UNIX_EPOCH},
};
use tokio::sync::Mutex;
#[derive(Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ThreadGoal {
    goal_id: String,
    thread_id: String,
    objective: String,
    status: String,
    token_budget: Option<u64>,
    tokens_used: u64,
    time_used_seconds: f64,
    created_at: u64,
    updated_at: u64,
}
#[derive(Default, Serialize, Deserialize)]
struct Persisted {
    goal: Option<ThreadGoal>,
    usage: BTreeMap<String, (u64, f64)>,
}
/// Durable JSON adapter for a single native thread. Keep one instance per thread;
/// call account_turn with cumulative uncached-input + output usage snapshots.
/// The caller owns continuation scheduling and explicit user resume controls.
pub struct FileGoalStore {
    path: PathBuf,
    thread_id: String,
    lock: Mutex<()>,
}
impl FileGoalStore {
    pub fn new(path: PathBuf, thread_id: String) -> Self {
        Self {
            path,
            thread_id,
            lock: Mutex::new(()),
        }
    }
    async fn load(&self) -> Result<Persisted, crate::contract::ToolError> {
        match tokio::fs::read(&self.path).await {
            Ok(bytes) => {
                let value: Persisted = serde_json::from_slice(&bytes)?;
                if value
                    .goal
                    .as_ref()
                    .is_some_and(|g| g.thread_id != self.thread_id)
                {
                    return Err(error("goal belongs to another thread"));
                }
                Ok(value)
            }
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Persisted::default()),
            Err(e) => Err(e.into()),
        }
    }
    async fn save(&self, value: &Persisted) -> Result<(), crate::contract::ToolError> {
        use tokio::io::AsyncWriteExt;
        if let Some(parent) = self.path.parent() {
            tokio::fs::create_dir_all(parent).await?;
        }
        let temporary = self
            .path
            .with_extension(format!("{}.tmp", uuid::Uuid::new_v4()));
        let mut options = tokio::fs::OpenOptions::new();
        options.write(true).create_new(true);
        #[cfg(unix)]
        options.mode(0o600);
        let mut file = options.open(&temporary).await?;
        file.write_all(&serde_json::to_vec(value)?).await?;
        file.sync_all().await?;
        drop(file);
        tokio::fs::rename(&temporary, &self.path).await?;
        Ok(())
    }
    fn bind(&self, context: ToolContext<'_>) -> Result<(), crate::contract::ToolError> {
        if context.session_id() != self.thread_id {
            return Err(error("goal context belongs to another thread"));
        }
        Ok(())
    }
    /// Idempotent cumulative accounting, bounded to the admitted goal identity.
    pub async fn account_turn(
        &self,
        goal_id: &str,
        turn: &str,
        tokens: u64,
        seconds: f64,
    ) -> Result<(), crate::contract::ToolError> {
        if !seconds.is_finite() || seconds < 0.0 {
            return Err(error("invalid goal usage"));
        }
        let _guard = self.lock.lock().await;
        let mut state = self.load().await?;
        let Some(goal) = &mut state.goal else {
            return Ok(());
        };
        if goal.goal_id != goal_id {
            return Ok(());
        }
        let previous = state.usage.get(turn).copied().unwrap_or((0, 0.0));
        let next = (tokens.max(previous.0), seconds.max(previous.1));
        goal.tokens_used = goal
            .tokens_used
            .checked_add(next.0 - previous.0)
            .ok_or_else(|| error("goal usage overflow"))?;
        goal.time_used_seconds += next.1 - previous.1;
        if !goal.time_used_seconds.is_finite() {
            return Err(error("goal usage overflow"));
        }
        if goal.token_budget.is_some_and(|b| goal.tokens_used >= b) && goal.status != "complete" {
            goal.status = "budgetLimited".into();
        }
        goal.updated_at = now();
        state.usage.insert(turn.to_owned(), next);
        self.save(&state).await
    }
}
fn now() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis()
        .try_into()
        .unwrap_or(u64::MAX)
}
fn response(goal: Option<&ThreadGoal>, complete: bool) -> Value {
    json!({ "goal": goal,
        "remainingTokens": goal.and_then(|g| g.token_budget.map(|b| b.saturating_sub(g.tokens_used))),
        "completionBudgetReport": goal.filter(|_| complete).and_then(|g| g.token_budget.map(|b| format!("Goal completed using {} of {b} budgeted tokens.", g.tokens_used))),
    })
}
#[async_trait::async_trait]
impl GoalStore for FileGoalStore {
    async fn get(&self, context: ToolContext<'_>) -> Result<Value, crate::contract::ToolError> {
        self.bind(context)?;
        let _guard = self.lock.lock().await;
        Ok(response(self.load().await?.goal.as_ref(), false))
    }
    async fn create(
        &self,
        objective: String,
        token_budget: Option<u64>,
        context: ToolContext<'_>,
    ) -> Result<Value, crate::contract::ToolError> {
        self.bind(context)?;
        let _guard = self.lock.lock().await;
        let mut state = self.load().await?;
        if state.goal.as_ref().is_some_and(|g| g.status != "complete") {
            return Err(error(
                "cannot create a new goal because this thread has an unfinished goal",
            ));
        }
        let timestamp = now();
        state.goal = Some(ThreadGoal {
            goal_id: uuid::Uuid::new_v4().to_string(),
            thread_id: self.thread_id.clone(),
            objective,
            token_budget,
            status: "active".into(),
            tokens_used: 0,
            time_used_seconds: 0.0,
            created_at: timestamp,
            updated_at: timestamp,
        });
        state.usage.clear();
        self.save(&state).await?;
        Ok(response(state.goal.as_ref(), false))
    }
    async fn update(
        &self,
        status: GoalStatus,
        context: ToolContext<'_>,
    ) -> Result<Value, crate::contract::ToolError> {
        self.bind(context)?;
        let _guard = self.lock.lock().await;
        let mut state = self.load().await?;
        let goal = state
            .goal
            .as_mut()
            .ok_or_else(|| error("this thread has no goal"))?;
        if !matches!(goal.status.as_str(), "budgetLimited" | "usageLimited") {
            goal.status = match status {
                GoalStatus::Complete => "complete",
                GoalStatus::Blocked => "blocked",
                GoalStatus::Paused => "paused",
            }
            .into();
            goal.updated_at = now();
        }
        let complete = goal.status == "complete";
        self.save(&state).await?;
        Ok(response(state.goal.as_ref(), complete))
    }
}
